### ISSUE TITLE
magit-clone create extra dir on cygwin / mingw

### ISSUE THREAD

**Issue by PierreTechoueyres** [magit/magit#2357]
`magit-clone` on cygwin create extra an directory named `c:` (or any letter mapped to the drive).

I think the directory used for the call of `magit-call-git` should be rewritten as a cygwin path.

I propose to add an `magit-shrink-git-file-name` for translating an Windows path to an cygwin one.
The change on `magit-expand-git-file-name` allow the translation of `~/`.  As `(file-name-absolute-p "~/")` is evaluated as `t` and so isn't translated.

``` patch
2 files changed, 10 insertions(+), 3 deletions(-)
lisp/magit-git.el    | 11 +++++++++--
lisp/magit-remote.el |  2 +-

modified   lisp/magit-git.el
@@ -534,14 +534,21 @@ Sorted from longest to shortest CYGWIN name."
   :type '(alist :key-type string :value-type directory))
 
 (defun magit-expand-git-file-name (filename)
-  (unless (file-name-absolute-p filename)
-    (setq filename (expand-file-name filename)))
+  (setq filename (expand-file-name filename))
   (-if-let ((cyg . win)
             (cl-assoc filename magit-cygwin-mount-points
                       :test (lambda (f cyg) (string-prefix-p cyg f))))
       (concat win (substring filename (length cyg)))
     filename))
 
+(defun magit-shrink-git-file-name (filename)
+  (setq filename (expand-file-name filename))
+  (-if-let ((cyg . win)
+            (cl-rassoc filename magit-cygwin-mount-points
+                       :test (lambda (f win) (string-prefix-p win f))))
+      (concat cyg (substring filename (length win)))
+    filename))
+
 (defun magit-decode-git-path (path)
   (if (eq (aref path 0) ?\")
       (string-as-multibyte (read path))
modified   lisp/magit-remote.el
@@ -47,7 +47,7 @@ Then show the status buffer for the new repository."
                        (match-string 1 url))))))))
   (make-directory directory t)
   (message "Cloning %s..." repository)
-  (when (= (magit-call-git "clone" repository directory) 0)
+  (when (= (magit-call-git "clone" repository (magit-shrink-git-file-name directory)) 0)
     (message "Cloning %s...done" repository)
     (magit-status-internal directory)))
```

**Comment by npostavs** [comment]
Can you show the exact call needed to produce this funny "c:" directory? I noticed I did actually have such a thing, so I must have triggered this before without noticing, but I couldn't make it happen again now.

**Comment by PierreTechoueyres** [comment]
Open the directory  `c:/tmp` in dired mode,
do a `magit-clone` for `https://github.com/magit/magit.git using c:/tmp/magit` as destination directory.
After the cloning you'll found an extra directory `c:` in your dired buffer (should be refreshed).

**Comment by npostavs** [comment]
Ah, even `git clone https://github.com/magit/magit.git c:/tmp/magit` from the cygwin command line does the same thing.

But why do you want to expand `~`? It seems like cygwin handles `~` correctly. And why you want to expand the file before "shrinking" it? As far as I can tell, cygwin handles relative pathnames correctly too.

**Comment by PierreTechoueyres** [comment]
I'm not certain to follow you for the "expand before "shrinking"" for relatives pathnames. And I guess that `shrink` isn't the right word in `magit-shrink-git-file-name`.

I think the directory path should be translated into a cygwin one before the call to git clone.
Maybe this should be extended to any call to a git command with an directory or file name with an absolute path in order to avoid the creation of an unwanted directory `c:`.

I think if git return  cygwin's paths they should be translated to Windows ones.

So I want to "expand" `~/` from Cygwin to the real Windows path. The home path for cygwin could be different from the Emac's home path (Yes you could think : this is stupid.)

**Comment by npostavs** [comment]
> I'm not certain to follow you for the "expand before "shrinking"" for relatives pathnames

I was trying to ask why you need the call to `expand-file-name` in `magit-shrink-git-file-name`:
> ```diff
> +(defun magit-shrink-git-file-name (filename) 
> +  (setq filename (expand-file-name filename))
> ```

In fact, as far as I can tell, cygwin can handle relative file names just fine, so perhaps we could make use of `file-relative-name`?

---

> The home path for cygwin could be different from the Emac's home path (Yes you could think : this is stupid.)

Well, it's a bad idea for sure. One major problem is that when you call cygwin programs from Emacs they will see the `HOME` env variable that Emacs has set, so the cygwin home path from Emacs will be different from the normal cygwin home path.

See also https://github.com/magit/magit/wiki/Pushing-with-Magit-from-Windows#before-starting-set-home

Basically, I strongly oppose supporting a multi `HOME` setup, it's too difficult to get right.

**Comment by PierreTechoueyres** [comment]
> I was trying to ask why you need the call to `expand-file-name` in `magit-shrink-git-file-name`

I just do the same thing as in `magit-expand-file-name` but I didn't test if it could work without this.

I'm not certain `file-relative-name` could be used for cloning because you could call `magit-clone` from a directory from a drive different of the one in the destination's path.
> Well, it's a bad idea for sure. One major problem is that when you call cygwin programs from Emacs they will see the HOME env variable that Emacs has set, so the cygwin home path from Emacs will be different from the normal cygwin home path.

I don't want to manage multiple home path. I just want it expands to an real Windows path.
If  `/home` is mounted on `d:/home` I want that `(magit-expand-git-file-name "~/tmp")` return  `d:/home/my_user/tmp`.

**Comment by npostavs** [comment]
> I think if git return cygwin's paths they should be translated to Windows ones.
> 
> So I want to "expand" ~/ from Cygwin to the real Windows path.

But git will never return a path with `~` in it (I think?).

> I just do the same thing as in magit-expand-file-name but I didn't test if it could work without this.

I think it should be okay to leave it relative, because cygwin doesn't have a problem with relative paths.


> I don't want to manage multiple home path. I just want it expands to an real Windows path.
> If /home is mounted on d:/home I want that (magit-expand-git-file-name "~/tmp") return d:/home/my_user/tmp.

Okay, but note this is only workable if Emacs and cygwin agree on the meaning of `~`.

**Comment by tarsius** [comment]
Git doesn't understand an incoming "~/" so I would expect it also never prints one. Because git doesn't understand "~/", `expand-file-name` should indeed always be used on all absolute paths passed to git. But relative paths should be kept that way because then potential disagreements about where $HOME is don't matter. Such disagreement still matters in other places, and I second the opinion that the user is responsible for resolving such conflicts.

> After the cloning you'll found an extra directory `c:` in your dired buffer (should be refreshed).

Do I understand this correctly, you end up with:

```
c:/tmp/magit/.git
c:/tmp/c:
```

as opposed to, say, `c:/tmp/c:/magit/.git`? If it's the former, then that might be related to #2354.

**Comment by PierreTechoueyres** [comment]
@tarsius Yes I end up with:
```
c:/tmp/magit/.git
c/tmp/c:
```
But I have no clues for #2354. May be you should try it and check what's done.
At least, the problem in `git-clone` is solved by `magit-shrink-git-file-name`.

@npostavs I may be wrong with `(magit-expand-git-file-name "~/tmp")`.
But, as you said, if git doesn't understand `~/` it must be translated in `magit-shrink-git-file-name`

**Comment by npostavs** [comment]
> If it's the former, then that might be related to #2354.

Hmm, I don't see the connection. 

From a few more experiments, the `c:` doesn't need to end up next to the cloned repository, the result is
```
c:/tmp/magit/.git
<cwd>/c:
```
where `<cwd>` is whatever directory `git clone` is called from, even if it's on another drive (and note that it's actually `c`, which cygwin recognizes as an encoding for `c:` (which would otherwise not be a legal filename)).

---

> But, as you said, if git doesn't understand ~/ it must be translated in magit-shrink-git-file-name

Yes. Fortunately, `file-relative-name` does do this:
`(file-relative-name "~/foo" "d:/bar") ;=> "c:/Users/npostavs/foo"`

Therefore I suggest:
```elisp
(defun magit-shrink-git-file-name (filename)
  (when (file-name-absolute-p filename)
    (setq filename (file-relative-name filename)))
  (-if-let ((cyg . win)
            (cl-rassoc filename magit-cygwin-mount-points
                       :test (lambda (f win) (string-prefix-p win f))))
      (concat cyg (substring filename (length win)))
    filename))
```
It has a pleasing symmetry vs `magit-expand-git-file-name`.

**Comment by tarsius** [comment]
Um, have you guys actually looked at `magit-clone`? I was struggling to figure out why git would create the repository but create some other directory - instead of say, just completely fail. Turns out that's not what happens, `magit-clone` itself does create the extra directory.

(Creating the directory with `make-directory` has the advantage that this allows creating a repository that is a subdirectory of a directory which previously did not exist. Don't know why I have added that though - probably someone asked for it.)

Does

```diff
@@ -45,7 +45,7 @@ (defun magit-clone (repository directory)
                   "Clone to: " nil nil nil
                   (and (string-match "\\([^./]+\\)\\(\\.git\\)?$" url)
                        (match-string 1 url))))))))
-  (make-directory directory t)
+  (make-directory (expand-file-name directory) t)
   (message "Cloning %s..." repository)
   (when (= (magit-call-git "clone" repository directory) 0)
     (message "Cloning %s...done" repository)
```

or something similar fix the issue?

**Comment by npostavs** [comment]
> I was struggling to imagining why git could possibly create the repository but also create some stray directory - instead of say, just completely fail. Turns out that's not what happens, magit-clone itself does create that extra directory.

No, no. As I said before, I see this extra directory being created when running clone from cygwin's command line, without invoking Emacs at all. Also it's not plausible that native win32 Emacs would produce a directory named `c`. The special unicode encoding for `:` is a cygwin thing. Here is a transcript:

```
/cygdrive/d/some-other-directory$ ls -a
.  ..
/cygdrive/d/some-other-directory$ ls -a /cygdrive/c/tmp/
.  ..
/cygdrive/d/some-other-directory$ git clone --depth=1 --single-branch https://github.com/magit/magit.git c:/tmp/magit
Cloning into 'c:/tmp/magit'...
remote: Counting objects: 68, done.
remote: Compressing objects: 100% (67/67), done.
remote: Total 68 (delta 3), reused 14 (delta 0), pack-reused 0
Unpacking objects: 100% (68/68), done.
Checking connectivity... done.
/cygdrive/d/some-other-directory$ ls -a
.  ..  c:
/cygdrive/d/some-other-directory$ ls -a /cygdrive/c/tmp/
.  ..  magit
/cygdrive/d/some-other-directory$
```
Looking the result from cmd.exe:
```
c:\tmp>dir d:\some-other-directory
 Volume in drive D is [...]
 Volume Serial Number is [...]

 Directory of d:\some-other-directory

25/10/2015  05:52 PM    <DIR>          .
25/10/2015  05:52 PM    <DIR>          ..
25/10/2015  05:52 PM    <DIR>          c?
               0 File(s)              0 bytes
               3 Dir(s)   2[...] bytes free
```

Oh, and turns out cygwin comes `strace`, here are the 1st few `mkdirs` from it:
```
   51   81213 [main] git 4600 mkdir: 0 = mkdir(c:, 511)
  406   82250 [main] git 4600 mkdir: 0 = mkdir(c:/tmp/magit, 511)
   79   98305 [main] git 4600 mkdir: 0 = mkdir(/cygdrive/c/tmp/magit/.git, 511)
   60  103444 [main] git 4600 mkdir: 0 = mkdir(/cygdrive/c/tmp/magit/.git/refs, 511)
```
